server/master/src/main/java/org/apache/accumulo/master/tableOps/bulkVer2/LoadFiles.java

```diff
@@ -328,38 +331,35 @@ private long loadFiles(Table.ID tableId, Path bulkDir, LoadMappingIterator lmi,
     loader.start(bulkDir, master, tid, bulkInfo.setTime);
 
     long t1 = System.currentTimeMillis();
-    while (true) {
-      if (loadMapEntry == null) {
-        if (!lmi.hasNext()) {
-          break;
-        }
-        loadMapEntry = lmi.next();
-      }
-      KeyExtent fke = loadMapEntry.getKey();
-      Files files = loadMapEntry.getValue();
-      loadMapEntry = null;
-
-      tablets.clear();
-
-      while (!Objects.equals(currentTablet.getPrevEndRow(), fke.getPrevEndRow())) {
-        currentTablet = tabletIter.next();
-      }
-      tablets.add(currentTablet);
+    while (lmi.hasNext()) {
+      loadMapEntry = lmi.next();
+      List<TabletMetadata> tablets = findOverlappingTablets(loadMapEntry.getKey(), tabletIter);
+      loader.load(tablets, loadMapEntry.getValue());
+    }
+    return loader.finish(Math.min(System.currentTimeMillis() - t1, 30000));
+  }
 
-      while (!Objects.equals(currentTablet.getEndRow(), fke.getEndRow())) {
-        currentTablet = tabletIter.next();
-        tablets.add(currentTablet);
-      }
+  /**
+   * Find all the tablets within the provided bulk load mapping range.
+   */
+  private List<TabletMetadata> findOverlappingTablets(KeyExtent loadRange,
+      Iterator<TabletMetadata> tabletIter) {
+    List<TabletMetadata> tablets = new ArrayList<>();
+    TabletMetadata currentTablet = tabletIter.next();
 
-      loader.load(tablets, files);
+    // skip tablets until we find the prevEndRow of loadRange
+    while (!Objects.equals(currentTablet.getPrevEndRow(), loadRange.getPrevEndRow())) {
+      currentTablet = tabletIter.next();
     }
-    long t2 = System.currentTimeMillis();
+    // we have found the first tablet in the range, add it to the list
+    tablets.add(currentTablet);
 
-    long sleepTime = loader.finish();
-    if (sleepTime > 0) {
```

Comment:
Personally I would leave this code as is in order to avoid duplicating the code for dealing with scan time.

Reply:
I guess I thought it was confusing that we had multiple calculations of the sleep time, one returned from the finish method and then more calculations dealing with the scan time.  I was trying to put it all in the one method (finish), even if it was in two places, offline and online.